{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f697b9ef-748d-4397-a6fb-c8b916b15068",
   "metadata": {},
   "outputs": [],
   "source": [
    "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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27b9f9c2-1b6a-4a78-baf5-2a728f633d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# once i have wholegenome_offset.trees\n",
    "import pandas as pd\n",
    "import tskit\n",
    "import allel\n",
    "import random\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import tsinfer\n",
    "import pyslim\n",
    "import os\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e98412d-752b-4bda-8963-7260bbb96779",
   "metadata": {},
   "outputs": [],
   "source": [
    "## create th einitital python file before the simulations, to set up the architectures, create variances, and optimas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd51083f-f215-4369-b9f3-20e45803d8d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c61571b-ae54-4c6e-9962-6de3a49f2683",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_vcf_offset = snakemake.params['og_vcf_offset'] \n",
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "\n",
    "pi_option =  snakemake.params['pi']\n",
    "pi = float(snakemake.params[pi_option])\n",
    "\n",
    "beta_option = snakemake.params['beta']\n",
    "beta = float(snakemake.params[pi_option])\n",
    "\n",
    "alelle_freq_option = snakemake.params['allele_freq']\n",
    "allele_freq = snakemake.params[alelle_freq_option]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "n_optima = str(snakemake.params['n_optima']) \n",
    "\n",
    "output_tree_seq_causalloci = snakemake.output[\"tree_seq_causalloci\"]\n",
    "output_loci_effectsize = snakemake.output[\"loci_effectsize\"]\n",
    "output_phenotypes = snakemake.output[\"phenotypes\"]\n",
    "output_optima_values = snakemake.output[\"optima_values\"]\n",
    "output_variance_values = snakemake.output[\"variance_values\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "e72137aa-16c8-4133-85eb-5e8250faa154",
   "metadata": {},
   "outputs": [],
   "source": [
    "## params hardcoded:\n",
    "og_vcf_offset = '../treeseq/wholegenome_offset.vcf'\n",
    "og_tree_offset = '../treeseq/wholegenome_offset_baselinetree.trees'\n",
    "\n",
    "pi = 20\n",
    "beta = 1\n",
    "\n",
    "allele_freq = [0.2, 0.3]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "n_optima = 5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc7d331-23db-412e-800d-46718b60fb72",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f5be3ff2-8fed-41f2-b5ce-9a464d0c4df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta):\n",
    "    alt_al_count = alt_al_per_pos.sum(axis=1)\n",
    "    alelle_dist = pd.DataFrame({'alt_al_count':alt_al_count, 'pos':pos})\n",
    "    alelle_dist['alt_al_freq'] = alelle_dist['alt_al_count'] / (n_ecotypes*2)\n",
    "    sim_freq_pos = alelle_dist[(alelle_dist['alt_al_freq'] < upper_bound) & (alelle_dist['alt_al_freq'] >= lower_bound)]['pos']\n",
    "    selected_sites = sim_freq_pos.sample(pi).values\n",
    "    sc = np.random.normal(0, beta, pi)\n",
    "    pos_sc = pd.DataFrame({'pos': selected_sites, 'sc': sc})\n",
    "    return pos_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "41e3f528-f5c9-49f6-82b1-00dac5187fab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_phenotypes(pos,pos_sc, alt_al_per_pos):\n",
    "    mask_positions = pd.Series(pos).isin(pos_sc['pos'])\n",
    "    alt_al_per_pos_selected_sites = alt_al_per_pos[mask_positions]\n",
    "    phenotypes = []\n",
    "    for i in range(alt_al_per_pos_selected_sites.shape[1]):\n",
    "        gen_effectsize = np.multiply(alt_al_per_pos_selected_sites[:, i] , pos_sc['sc'])\n",
    "        phenotypes.append(gen_effectsize.sum())\n",
    "    return phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "27ff97bf-802c-4b82-86a4-21e4df987407",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_optima(phenotypes):\n",
    "    max_pheno = max(phenotypes)\n",
    "    min_pheno = min(phenotypes)\n",
    "\n",
    "    length = max_pheno - min_pheno\n",
    "    step = length/(int(n_optima) - 1)\n",
    "    optima = [round(min_pheno + i * step, 4) for i in range(0, int(n_optima))]\n",
    "    return optima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2a915f1d-d8fe-475c-abaa-2aac7262a2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_variances(phenotypes, optima):\n",
    "    range_pheno =  max(phenotypes) - min(phenotypes)\n",
    "    dist_between_env = range_pheno / len(optima)\n",
    "    ## Strong selection, fitness 0 in the adyacent environemnt\n",
    "    sd1 = dist_between_env / 3\n",
    "    variance1 = sd1**2\n",
    "    ## moderate selection, fitness 0 in the other extremee environemnt \n",
    "    sd2 = (dist_between_env * 4) / 3 ## 3 sd will be in between 4  environments \n",
    "    variance2 = sd2**2\n",
    "    ## weak selection, half fitness in the other extreme environment \n",
    "    sd3 = (dist_between_env * 8) / 3\n",
    "    variance3 = sd3**2\n",
    "    variances = [sd1, sd2, sd3]\n",
    "    return variances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9f8051b8-edaa-4461-873a-4f256a94119d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc):\n",
    "    ts = tskit.load(og_tree_offset)\n",
    "\n",
    "    # dumpt the tables from the tree\n",
    "    tables = ts.dump_tables()\n",
    "\n",
    "    ## extract all teh sites from the og tree\n",
    "    complete_sites = pd.Series(tables.sites.position)\n",
    "\n",
    "    # create a mask to filter only the ones present in the selected sites (causal loci)\n",
    "    mask_delete_sites = complete_sites.isin(pos_sc['pos'])\n",
    "\n",
    "    ## replace the table only with the causal sites, and same for mutation tables\n",
    "    tables.sites.replace_with(tables.sites[mask_delete_sites])\n",
    "    tables.mutations.replace_with(tables.mutations[mask_delete_sites])\n",
    "    ## extract the new site index\n",
    "    tables.mutations.site = np.array(range(0, len(tables.mutations))).astype('int32')\n",
    "\n",
    "    ## ge tthe positions and sc in teh right order \n",
    "    pos_table = pd.Series(tables.sites.position).reset_index()\n",
    "    right_order_pos = pos_sc.merge(pos_table, left_on='pos',right_on =0).sort_values('index')\n",
    "\n",
    "    ## create the tree to then modify it \n",
    "    new_ts = tables.tree_sequence()\n",
    "    tables = new_ts.dump_tables()\n",
    "\n",
    "    ## chance the ancestral state to empty or slim will complain\n",
    "    tables.sites.clear()\n",
    "    for s in new_ts.sites():\n",
    "        tables.sites.append(s.replace(ancestral_state=\"\"))\n",
    "\n",
    "    ## add the selection coefficient and the rigth emtadata fro slim \n",
    "    tables.mutations.clear()\n",
    "    for k, (m, sc) in enumerate(zip(new_ts.mutations(), right_order_pos['sc'])):\n",
    "        mm = pyslim.default_slim_metadata('mutation_list_entry')\n",
    "        mm['selection_coeff'] = sc\n",
    "        tables.mutations.append(\n",
    "            m.replace(derived_state=str(k), metadata={'mutation_list': [mm]}))\n",
    "        \n",
    "    return tables.tree_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e90d1e9-9733-4119-b48b-1ae17d2e3b6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0eeb2213-d333-4165-b3d8-e55f9a866a1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for this im gonna use the og vcf file wth the offset to be able to map the positions correctly \n",
    "vcf_og = allel.read_vcf(og_vcf_offset, fields=[\"calldata/GT\", 'variants/POS' , 'samples'])\n",
    "geno_og = vcf_og[\"calldata/GT\"]\n",
    "samples = vcf_og['samples']\n",
    "pos = vcf_og['variants/POS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "35528e71-f55b-4947-bc34-58923a2f272f",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_ecotypes = len(vcf_og['samples'])\n",
    "alt_al_per_pos = geno_og.sum(axis=2) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "01e132a4-d3a4-41b2-bfc7-48344ba572b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_sc = calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0f5afc02-fc2b-4a42-b417-5cd55cfeb2eb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "phenotypes = calc_phenotypes(pos,pos_sc, alt_al_per_pos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "97fab941-9460-4082-b499-db1835ed0590",
   "metadata": {},
   "outputs": [],
   "source": [
    "optima = calc_optima(phenotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0622f532-7daa-4023-8fab-9a218912c2f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "variances = calc_variances(phenotypes, optima)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "b191f319-a91c-4449-bbb2-2ad515535bc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1.1716161941831824, 4.68646477673273, 9.37292955346546]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "fd461295-8c7e-45e6-9060-a93498e204f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(index =  ['strongsel','moderatesel','lowsel'],data = { 'var': variances}).to_csv('sel_var.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "3a17f44a-b143-4e23-8cda-1dae571d1712",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(data = {'selection': ['strongsel','moderatesel','lowsel'], 'var': variances}).to_csv('sel_var.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7225b0f6-63c1-48c7-af51-03560f57ae41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## save "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "65b89fda-d307-44c1-b29a-057637b0d895",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'output_phenotypes' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_19857/2398519311.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSeries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mphenotypes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_phenotypes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mpos_sc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_loci_effectsize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput_optima_values\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0melement\u001b[0m \u001b[0;32min\u001b[0m \u001b[0moptima\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'output_phenotypes' is not defined"
     ]
    }
   ],
   "source": [
    "pd.Series(phenotypes).to_csv(output_phenotypes)\n",
    "pos_sc.to_csv(output_loci_effectsize)\n",
    "\n",
    "with open(output_optima_values, 'w') as file:\n",
    "    for element in optima:\n",
    "        file.write(str(element) + '\\n')  # Write element followed by a newline\n",
    "\n",
    "with open(output_variance_values, 'w') as file:\n",
    "    for element in variances:\n",
    "        file.write(str(round(element,4)) + '\\n')  # Write element followed by a newline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8aa32c9-ea14-4c93-b217-5fde97e12d48",
   "metadata": {},
   "outputs": [],
   "source": [
    "### filter tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "96eea5eb-af0d-404c-a98d-4768d930bc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_slim_tree = keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6a2d94-ac2a-4c28-aa1d-1573bcab4906",
   "metadata": {},
   "outputs": [],
   "source": [
    "## save tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "a740f195-0aee-4c83-b6df-e54b24d92a7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "pre_slim_tree.dump(output_tree_seq_causalloci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1d73bd-9371-415b-9ee9-327506d4d64f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e6547a54-643f-4392-b93a-728f693b01f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## tree postprocesssing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f304969-0805-4a21-946a-83bcc370cfc8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f660b7-2c96-4370-ad40-d757f221d514",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "mapper_realid_metadataid = snakemake.input['mapper_ids'] \n",
    "output_sim_tree = snakemake.input['output_sim_tree'] \n",
    "output_sim_tree_wm = snakemake.output['output_sim_tree_wm'] \\\n",
    "output_vcf = snakemake.output['output_vcf'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "3cf82bfb-de0c-4541-8b69-e0ddacb7b761",
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap_neutral_mut (ts_new, ts, mapper_realid_metadataid):\n",
    "    ## extract surviving ndoes and comapre them to our old ndoes to place mtuations in the right place\n",
    "    surviving_nodes = []\n",
    "    for i in ts_new.tables.nodes:\n",
    "        surviving_nodes.append(i.metadata['slim_id'])\n",
    "    ## new nodes id and the ids i gave them in the past\n",
    "    new_mapper = pd.DataFrame({'new_ids': range(0, len(ts_new.tables.nodes)), 'my_ids_metadata':surviving_nodes})\n",
    "    ## map old nodes with new nodes\n",
    "    mapper_lost_nodes = new_mapper.merge(mapper_realid_metadataid, left_on = 'my_ids_metadata', right_on = 'my_ids_metadata', how= 'right')\n",
    "\n",
    "    ## create a mask to only keep from the old nodes the ones that survived the simulation\n",
    "    mask = mapper_lost_nodes['new_ids'].notna()\n",
    "\n",
    "    tables_og = ts.dump_tables()\n",
    "\n",
    "    ## now filter old tables only based on surviving nodes \n",
    "    tables_og.nodes.replace_with(tables_og.nodes[mask])\n",
    "\n",
    "    ## now filter mutation table based on the surviving nodes, for that, extract the nodes \n",
    "    old_nodes = tables_og.mutations.node\n",
    "\n",
    "    old_nodes = pd.Series(old_nodes)\n",
    "\n",
    "    old_nodes.name = 'old_nodes'\n",
    "\n",
    "    ## create a dataframe relating the new and old nodes\n",
    "    replace_oldbynew_nodes = pd.merge(old_nodes, mapper_lost_nodes, left_on ='old_nodes', right_on = 'real_id', how= 'left')\n",
    "\n",
    "    ## create a mask to filter out all the mutations than has been lost \n",
    "    mask_mutations_lost = replace_oldbynew_nodes['new_ids'].notna()\n",
    "\n",
    "    ## filter out mutations that has been lost \n",
    "    table_mutations = tables_og.mutations[mask_mutations_lost]\n",
    "\n",
    "    ## now replace the old nodes ids by the new nodes ids with the mapper\n",
    "    ids_to_replace = replace_oldbynew_nodes.dropna()['new_ids']\n",
    "    table_mutations.node = np.array(ids_to_replace.astype('int32'))\n",
    "\n",
    "    ## and jsut set the sites from 0 to the length of mutation table \n",
    "    table_mutations.site = np.array(range(0, len(table_mutations))).astype('int32')\n",
    "\n",
    "    ## apply the same filter from the mutations table to the sites table \n",
    "    table_sites = tables_og.sites[mask_mutations_lost]  \n",
    "\n",
    "    ## now replace all this filter old tables in the new tree seq! \n",
    "    new_tables = ts_new.dump_tables()\n",
    "\n",
    "    new_tables.mutations.replace_with(table_mutations)\n",
    "\n",
    "    new_tables.sites.replace_with(table_sites)\n",
    "\n",
    "    ## make sure to compure mutations parents\n",
    "    new_tables.compute_mutation_parents()\n",
    "\n",
    "    ## create tree seq based on tables\n",
    "    tree_nm = new_tables.tree_sequence()\n",
    "\n",
    "    return tree_nm.simplify()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "0848ee38-29f5-40da-868e-e118cfe75ae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_tree_to_vcf (tree,name_vcf):\n",
    "    # create a vcf file from the treeseq \n",
    "    with open(name_vcf, 'w') as file:\n",
    "        # Pass the file object as the output parameter\n",
    "        tree.write_vcf(output=file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "213857cb-4d0d-45af-87be-2acce7b7fa4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import the old tree\n",
    "ts_old = tskit.load(og_tree_offset)\n",
    "#import mapper old nodes to new nodes\n",
    "mapper_realid_metadataid = pd.read_csv('../treeseq/mapper_realid_metadataid_wholegenome.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64866d9c-8e66-405d-8361-7981a4e96295",
   "metadata": {},
   "outputs": [],
   "source": [
    "## ts new\n",
    "ts_new = tskit.load(output_sim_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "3c0b15bd-311d-40a0-9011-39f1036af7d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <div>\n",
       "            <style scoped=\"\">\n",
       "                .tskit-table tbody tr th:only-of-type {vertical-align: middle;}\n",
       "                .tskit-table tbody tr th {vertical-align: top;}\n",
       "                .tskit-table tbody td {text-align: right;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody th {padding: 0.5em 0.5em;}\n",
       "            </style>\n",
       "            <table border=\"1\" class=\"tskit-table\">\n",
       "                <thead>\n",
       "                    <tr>\n",
       "                        <th>id</th><th>flags</th><th>population</th><th>individual</th><th>time</th><th>metadata</th>\n",
       "                    </tr>\n",
       "                </thead>\n",
       "                <tbody>\n",
       "                    <tr><td>0</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td>{'slim_id': 0, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>1</td><td>1</td><td>0</td><td>0</td><td>0.00000000</td><td>{'slim_id': 1, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>2</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td>{'slim_id': 2, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>3</td><td>1</td><td>0</td><td>1</td><td>0.00000000</td><td>{'slim_id': 3, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>4</td><td>1</td><td>0</td><td>2</td><td>0.00000000</td><td>{'slim_id': 4, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>5</td><td>1</td><td>0</td><td>2</td><td>0.00000000</td><td>{'slim_id': 5, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>6</td><td>1</td><td>0</td><td>3</td><td>0.00000000</td><td>{'slim_id': 6, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>7</td><td>1</td><td>0</td><td>3</td><td>0.00000000</td><td>{'slim_id': 7, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>8</td><td>1</td><td>0</td><td>4</td><td>0.00000000</td><td>{'slim_id': 8, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>9</td><td>1</td><td>0</td><td>4</td><td>0.00000000</td><td>{'slim_id': 9, 'is_null': False, 'gen...</td></tr>\n",
       "<tr><td>10</td><td>1</td><td>0</td><td>5</td><td>0.00000000</td><td>{'slim_id': 10, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>11</td><td>1</td><td>0</td><td>5</td><td>0.00000000</td><td>{'slim_id': 11, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>12</td><td>1</td><td>0</td><td>6</td><td>0.00000000</td><td>{'slim_id': 12, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>13</td><td>1</td><td>0</td><td>6</td><td>0.00000000</td><td>{'slim_id': 13, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>14</td><td>1</td><td>0</td><td>7</td><td>0.00000000</td><td>{'slim_id': 14, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>15</td><td>1</td><td>0</td><td>7</td><td>0.00000000</td><td>{'slim_id': 15, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>16</td><td>1</td><td>0</td><td>8</td><td>0.00000000</td><td>{'slim_id': 16, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>17</td><td>1</td><td>0</td><td>8</td><td>0.00000000</td><td>{'slim_id': 17, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>18</td><td>1</td><td>0</td><td>9</td><td>0.00000000</td><td>{'slim_id': 18, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td>19</td><td>1</td><td>0</td><td>9</td><td>0.00000000</td><td>{'slim_id': 19, 'is_null': False, 'ge...</td></tr>\n",
       "<tr><td style=\"text-align: center;\" colspan=\"93\"><em>2146124 rows skipped (tskit.set_print_options)</em></td></tr>\n",
       "<tr><td>2146144</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145683, 'is_null': True...</td></tr>\n",
       "<tr><td>2146145</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145684, 'is_null': True...</td></tr>\n",
       "<tr><td>2146146</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145685, 'is_null': True...</td></tr>\n",
       "<tr><td>2146147</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145686, 'is_null': True...</td></tr>\n",
       "<tr><td>2146148</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145687, 'is_null': True...</td></tr>\n",
       "<tr><td>2146149</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145688, 'is_null': True...</td></tr>\n",
       "<tr><td>2146150</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145689, 'is_null': True...</td></tr>\n",
       "<tr><td>2146151</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145690, 'is_null': True...</td></tr>\n",
       "<tr><td>2146152</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145691, 'is_null': True...</td></tr>\n",
       "<tr><td>2146153</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145692, 'is_null': True...</td></tr>\n",
       "<tr><td>2146154</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145693, 'is_null': True...</td></tr>\n",
       "<tr><td>2146155</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145694, 'is_null': True...</td></tr>\n",
       "<tr><td>2146156</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145695, 'is_null': True...</td></tr>\n",
       "<tr><td>2146157</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145696, 'is_null': True...</td></tr>\n",
       "<tr><td>2146158</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145697, 'is_null': True...</td></tr>\n",
       "<tr><td>2146159</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145698, 'is_null': True...</td></tr>\n",
       "<tr><td>2146160</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145699, 'is_null': True...</td></tr>\n",
       "<tr><td>2146161</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145700, 'is_null': True...</td></tr>\n",
       "<tr><td>2146162</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145701, 'is_null': True...</td></tr>\n",
       "<tr><td>2146163</td><td>0</td><td>-1</td><td>-1</td><td>0.99809946</td><td>{'slim_id': -2145702, 'is_null': True...</td></tr>\n",
       "\n",
       "                </tbody>\n",
       "            </table>\n",
       "        </div>\n",
       "    "
      ],
      "text/plain": [
       "<tskit.tables.NodeTable at 0x7fcbe4927070>"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts_nm = overlap_neutral_mut(ts_new, ts_old, mapper_realid_metadataid)\n",
    "tskit.dump(ts_nm, output_sim_tree_wm)\n",
    "convert_tree_to_vcf(ts_nm, output_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd913e4e-2e7b-4947-a367-0e6c62fe6e19",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
