{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f697b9ef-748d-4397-a6fb-c8b916b15068",
   "metadata": {},
   "outputs": [],
   "source": [
    "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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "27b9f9c2-1b6a-4a78-baf5-2a728f633d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# once i have wholegenome_offset.trees\n",
    "import pandas as pd\n",
    "import tskit\n",
    "import allel\n",
    "import random\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import tsinfer\n",
    "import pyslim\n",
    "import os\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e98412d-752b-4bda-8963-7260bbb96779",
   "metadata": {},
   "outputs": [],
   "source": [
    "## create th einitital python file before the simulations, to set up the architectures, create variances, and optimas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd51083f-f215-4369-b9f3-20e45803d8d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c61571b-ae54-4c6e-9962-6de3a49f2683",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_vcf_offset = snakemake.input['og_vcf_offset'] \n",
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "\n",
    "pi_option =  snakemake.params['pi']\n",
    "pi = int(snakemake.params[pi_option])\n",
    "\n",
    "beta_option = snakemake.params['beta']\n",
    "beta = int(snakemake.params[beta_option])\n",
    "\n",
    "alelle_freq_option = snakemake.params['allele_freq']\n",
    "allele_freq = snakemake.params[alelle_freq_option]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "optima_qty = str(snakemake.params['optima_qty']) \n",
    "\n",
    "output_tree_seq_causalloci = snakemake.output[\"tree_seq_causalloci\"]\n",
    "output_loci_effectsize = snakemake.output[\"loci_effectsize\"]\n",
    "output_phenotypes = snakemake.output[\"phenotypes\"]\n",
    "output_optima_values = snakemake.output[\"optima_values\"]\n",
    "output_variance_values = snakemake.output[\"variance_values\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e72137aa-16c8-4133-85eb-5e8250faa154",
   "metadata": {},
   "outputs": [],
   "source": [
    "## params hardcoded:\n",
    "og_vcf_offset = '../treeseq/wholegenome_offset.vcf'\n",
    "og_tree_offset = '../treeseq/wholegenome_offset_baselinetree.trees'\n",
    "\n",
    "pi = 20\n",
    "beta = 1\n",
    "\n",
    "allele_freq = [0.2, 0.3]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "optima_qty = 5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1dc7d331-23db-412e-800d-46718b60fb72",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "125fdca8-815d-41b0-b2e9-6b53694c1214",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta):\n",
    "    alt_al_count = alt_al_per_pos.sum(axis=1)\n",
    "    alelle_dist = pd.DataFrame({'alt_al_count':alt_al_count, 'pos':pos})\n",
    "    alelle_dist['alt_al_freq'] = alelle_dist['alt_al_count'] / (n_ecotypes*2)\n",
    "    sim_freq_pos = alelle_dist[(alelle_dist['alt_al_freq'] < upper_bound) & (alelle_dist['alt_al_freq'] >= lower_bound)]['pos']\n",
    "    selected_sites = sim_freq_pos.sample(pi).values\n",
    "    sc = np.random.normal(0, beta, pi)\n",
    "    pos_sc = pd.DataFrame({'pos': selected_sites, 'sc': sc})\n",
    "    return pos_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "b9f0cfa3-2dd2-48ac-86a4-8ef70d3d6e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_phenotypes(pos,pos_sc, alt_al_per_pos):\n",
    "    mask_positions = pd.Series(pos).isin(pos_sc['pos'])\n",
    "    alt_al_per_pos_selected_sites = alt_al_per_pos[mask_positions]\n",
    "    phenotypes = []\n",
    "    for i in range(alt_al_per_pos_selected_sites.shape[1]):\n",
    "        gen_effectsize = np.multiply(alt_al_per_pos_selected_sites[:, i] , pos_sc['sc'])\n",
    "        phenotypes.append(gen_effectsize.sum())\n",
    "    return phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "981b17fc-83c9-4581-96ab-4cc6d287d530",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_optima(phenotypes):\n",
    "    max_pheno = max(phenotypes)\n",
    "    min_pheno = min(phenotypes)\n",
    "\n",
    "    length = max_pheno - min_pheno\n",
    "    step = length/(int(optima_qty) - 1)\n",
    "    optima = [round(min_pheno + i * step, 4) for i in range(0, int(optima_qty))]\n",
    "    return optima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "17bf89d2-0352-4a9f-917d-977679d7d945",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_variances(phenotypes, optima):\n",
    "    range_pheno =  max(phenotypes) - min(phenotypes)\n",
    "    dist_between_env = range_pheno / len(optima)\n",
    "    ## Strong selection, fitness 0 in the adyacent environemnt\n",
    "    sd1 = dist_between_env / 3\n",
    "    variance1 = sd1**2\n",
    "    ## moderate selection, fitness 0 in the other extremee environemnt \n",
    "    sd2 = (dist_between_env * 4) / 3 ## 3 sd will be in between 4  environments \n",
    "    variance2 = sd2**2\n",
    "    ## weak selection, half fitness in the other extreme environment \n",
    "    sd3 = (dist_between_env * 8) / 3\n",
    "    variance3 = sd3**2\n",
    "    variances = [sd1, sd2, sd3]\n",
    "    return variances\n",
    "\n",
    "def keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc):\n",
    "    ts = tskit.load(og_tree_offset)\n",
    "\n",
    "    # dumpt the tables from the tree\n",
    "    tables = ts.dump_tables()\n",
    "\n",
    "    ## extract all teh sites from the og tree\n",
    "    complete_sites = pd.Series(tables.sites.position)\n",
    "\n",
    "    # create a mask to filter only the ones present in the selected sites (causal loci)\n",
    "    mask_delete_sites = complete_sites.isin(pos_sc['pos'])\n",
    "\n",
    "    ## replace the table only with the causal sites, and same for mutation tables\n",
    "    tables.sites.replace_with(tables.sites[mask_delete_sites])\n",
    "    tables.mutations.replace_with(tables.mutations[mask_delete_sites])\n",
    "    ## extract the new site index\n",
    "    tables.mutations.site = np.array(range(0, len(tables.mutations))).astype('int32')\n",
    "\n",
    "    ## ge tthe positions and sc in teh right order \n",
    "    pos_table = pd.Series(tables.sites.position).reset_index()\n",
    "    right_order_pos = pos_sc.merge(pos_table, left_on='pos',right_on =0).sort_values('index')\n",
    "\n",
    "    ## create the tree to then modify it \n",
    "    new_ts = tables.tree_sequence()\n",
    "    tables = new_ts.dump_tables()\n",
    "\n",
    "    ## chance the ancestral state to empty or slim will complain\n",
    "    tables.sites.clear()\n",
    "    for s in new_ts.sites():\n",
    "        tables.sites.append(s.replace(ancestral_state=\"\"))\n",
    "\n",
    "    ## add the selection coefficient and the rigth emtadata fro slim \n",
    "    tables.mutations.clear()\n",
    "    for k, (m, sc) in enumerate(zip(new_ts.mutations(), right_order_pos['sc'])):\n",
    "        mm = pyslim.default_slim_metadata('mutation_list_entry')\n",
    "        mm['selection_coeff'] = sc\n",
    "        tables.mutations.append(\n",
    "            m.replace(derived_state=str(k), metadata={'mutation_list': [mm]}))\n",
    "        \n",
    "    return tables.tree_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "1e90d1e9-9733-4119-b48b-1ae17d2e3b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for this im gonna use the og vcf file wth the offset to be able to map the positions correctly \n",
    "vcf_og = allel.read_vcf(og_vcf_offset, fields=[\"calldata/GT\", 'variants/POS' , 'samples'])\n",
    "geno_og = vcf_og[\"calldata/GT\"]\n",
    "samples = vcf_og['samples']\n",
    "pos = vcf_og['variants/POS']\n",
    "\n",
    "n_ecotypes = len(vcf_og['samples'])\n",
    "alt_al_per_pos = geno_og.sum(axis=2) \n",
    "\n",
    "pos_sc = calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16706eb9-38fb-4d1f-8a5d-7714811dcc48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "f1996aab-1c7c-42d5-b351-c33a9a1d6aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "phenotypes = calc_phenotypes(pos,pos_sc, alt_al_per_pos)\n",
    "\n",
    "optima = calc_optima(phenotypes)\n",
    "\n",
    "variances = calc_variances(phenotypes, optima)\n",
    "\n",
    "variances\n",
    "\n",
    "pd.DataFrame(index =  ['strongsel','moderatesel','lowsel'],data = { 'var': variances}).to_csv('sel_var.csv')\n",
    "\n",
    "pd.DataFrame(data = {'selection': ['strongsel','moderatesel','lowsel'], 'var': variances}).to_csv('sel_var.csv')\n",
    "\n",
    "## save \n",
    "\n",
    "pd.Series(phenotypes).to_csv(output_phenotypes)\n",
    "pos_sc.to_csv(output_loci_effectsize)\n",
    "\n",
    "with open(output_optima_values, 'w') as file:\n",
    "    for element in optima:\n",
    "        file.write(str(element) + '\\n')  # Write element followed by a newline\n",
    "\n",
    "with open(output_variance_values, 'w') as file:\n",
    "    for element in variances:\n",
    "        file.write(str(round(element,4)) + '\\n')  # Write element followed by a newline\n",
    "\n",
    "### filter tree\n",
    "\n",
    "pre_slim_tree = keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc)\n",
    "\n",
    "## save tree\n",
    "\n",
    "pre_slim_tree.dump(output_tree_seq_causalloci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1d73bd-9371-415b-9ee9-327506d4d64f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e6547a54-643f-4392-b93a-728f693b01f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## tree postprocesssing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f304969-0805-4a21-946a-83bcc370cfc8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "63c188e0-249a-4975-81ae-8ac0fbfb0a0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "mapper_realid_metadataid = snakemake.input['mapper_ids'] \n",
    "output_sim_tree = snakemake.input['output_sim_tree'] \n",
    "output_sim_tree_wm = snakemake.output['output_sim_tree_wm'] \\\n",
    "output_vcf = snakemake.output['output_vcf'] \n",
    "\n",
    "def overlap_neutral_mut (ts_new, ts, mapper_realid_metadataid):\n",
    "    ## extract surviving ndoes and comapre them to our old ndoes to place mtuations in the right place\n",
    "    surviving_nodes = []\n",
    "    for i in ts_new.tables.nodes:\n",
    "        surviving_nodes.append(i.metadata['slim_id'])\n",
    "    ## new nodes id and the ids i gave them in the past\n",
    "    new_mapper = pd.DataFrame({'new_ids': range(0, len(ts_new.tables.nodes)), 'my_ids_metadata':surviving_nodes})\n",
    "    ## map old nodes with new nodes\n",
    "    mapper_lost_nodes = new_mapper.merge(mapper_realid_metadataid, left_on = 'my_ids_metadata', right_on = 'my_ids_metadata', how= 'right')\n",
    "\n",
    "    ## create a mask to only keep from the old nodes the ones that survived the simulation\n",
    "    mask = mapper_lost_nodes['new_ids'].notna()\n",
    "\n",
    "    tables_og = ts.dump_tables()\n",
    "\n",
    "    ## now filter old tables only based on surviving nodes \n",
    "    tables_og.nodes.replace_with(tables_og.nodes[mask])\n",
    "\n",
    "    ## now filter mutation table based on the surviving nodes, for that, extract the nodes \n",
    "    old_nodes = tables_og.mutations.node\n",
    "\n",
    "    old_nodes = pd.Series(old_nodes)\n",
    "\n",
    "    old_nodes.name = 'old_nodes'\n",
    "\n",
    "    ## create a dataframe relating the new and old nodes\n",
    "    replace_oldbynew_nodes = pd.merge(old_nodes, mapper_lost_nodes, left_on ='old_nodes', right_on = 'real_id', how= 'left')\n",
    "\n",
    "    ## create a mask to filter out all the mutations than has been lost \n",
    "    mask_mutations_lost = replace_oldbynew_nodes['new_ids'].notna()\n",
    "\n",
    "    ## filter out mutations that has been lost \n",
    "    table_mutations = tables_og.mutations[mask_mutations_lost]\n",
    "\n",
    "    ## now replace the old nodes ids by the new nodes ids with the mapper\n",
    "    ids_to_replace = replace_oldbynew_nodes.dropna()['new_ids']\n",
    "    table_mutations.node = np.array(ids_to_replace.astype('int32'))\n",
    "\n",
    "    ## and jsut set the sites from 0 to the length of mutation table \n",
    "    table_mutations.site = np.array(range(0, len(table_mutations))).astype('int32')\n",
    "\n",
    "    ## apply the same filter from the mutations table to the sites table \n",
    "    table_sites = tables_og.sites[mask_mutations_lost]  \n",
    "\n",
    "    ## now replace all this filter old tables in the new tree seq! \n",
    "    new_tables = ts_new.dump_tables()\n",
    "\n",
    "    new_tables.mutations.replace_with(table_mutations)\n",
    "\n",
    "    new_tables.sites.replace_with(table_sites)\n",
    "\n",
    "    ## make sure to compure mutations parents\n",
    "    new_tables.compute_mutation_parents()\n",
    "\n",
    "    ## create tree seq based on tables\n",
    "    tree_nm = new_tables.tree_sequence()\n",
    "\n",
    "    return tree_nm.simplify()\n",
    "\n",
    "def convert_tree_to_vcf (tree,name_vcf):\n",
    "    # create a vcf file from the treeseq \n",
    "    with open(name_vcf, 'w') as file:\n",
    "        # Pass the file object as the output parameter\n",
    "        tree.write_vcf(output=file)\n",
    "\n",
    "#import the old tree\n",
    "ts_old = tskit.load(og_tree_offset)\n",
    "#import mapper old nodes to new nodes\n",
    "mapper_realid_metadataid = pd.read_csv('../treeseq/mapper_realid_metadataid_wholegenome.csv')\n",
    "\n",
    "## ts new\n",
    "ts_new = tskit.load(output_sim_tree)\n",
    "\n",
    "ts_nm = overlap_neutral_mut(ts_new, ts_old, mapper_realid_metadataid)\n",
    "tskit.dump(ts_nm, output_sim_tree_wm)\n",
    "convert_tree_to_vcf(ts_nm, output_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b1244e-8a50-4c97-a1e6-9aa6feb41844",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5fdd396-fdbb-4bd5-b0d3-d74bd3fdac96",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "c7bd449c-7ce7-4aa0-8b79-3e0c1bab42f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.167066, 0.238523, 0.000196125, 0.991214, 0.00760604, 2.5717e-05, 0.144752, 0.321249, 0.637912, 0.858058, 6.23198e-06, 0.119282, 0.111827, 1.95253e-05, 0.000294039, 2.36393e-16, 0.00054201, 5.51868e-17, 0.000307715, 2.2513e-05, 0.0189577, 0.181709, 1.53508e-06, 0.0124593, 4.06938e-08, 0.580586, 0.00320625, 0.00136106, 6.36069e-06, 0.133209, 0.86022, 0.000208478, 1.6127e-14, 0.000654452]\n"
     ]
    }
   ],
   "source": [
    "input_values = [\n",
    "    \"fitness\",\n",
    "    0.167066,\n",
    "    \"fitness\",\n",
    "    0.238523,\n",
    "    \"fitness\",\n",
    "    0.000196125,\n",
    "    \"fitness\",\n",
    "    0.991214,\n",
    "    \"fitness\",\n",
    "    0.00760604,\n",
    "    \"fitness\",\n",
    "    2.5717e-05,\n",
    "    \"fitness\",\n",
    "    0.144752,\n",
    "    \"fitness\",\n",
    "    0.321249,\n",
    "    \"fitness\",\n",
    "    0.637912,\n",
    "    \"fitness\",\n",
    "    0.858058,\n",
    "    \"fitness\",\n",
    "    6.23198e-06,\n",
    "    \"fitness\",\n",
    "    0.119282,\n",
    "    \"fitness\",\n",
    "    0.111827,\n",
    "    \"fitness\",\n",
    "    1.95253e-05,\n",
    "    \"fitness\",\n",
    "    0.000294039,\n",
    "    \"fitness\",\n",
    "    2.36393e-16,\n",
    "    \"fitness\",\n",
    "    0.00054201,\n",
    "    \"fitness\",\n",
    "    5.51868e-17,\n",
    "    \"fitness\",\n",
    "    0.000307715,\n",
    "    \"fitness\",\n",
    "    2.2513e-05,\n",
    "    \"fitness\",\n",
    "    0.0189577,\n",
    "    \"fitness\",\n",
    "    0.181709,\n",
    "    \"fitness\",\n",
    "    1.53508e-06,\n",
    "    \"fitness\",\n",
    "    0.0124593,\n",
    "    \"fitness\",\n",
    "    4.06938e-08,\n",
    "    \"fitness\",\n",
    "    0.580586,\n",
    "    \"fitness\",\n",
    "    0.00320625,\n",
    "    \"fitness\",\n",
    "    0.00136106,\n",
    "    \"fitness\",\n",
    "    6.36069e-06,\n",
    "    \"fitness\",\n",
    "    0.133209,\n",
    "    \"fitness\",\n",
    "    0.86022,\n",
    "    \"fitness\",\n",
    "    0.000208478,\n",
    "    \"fitness\",\n",
    "    1.6127e-14,\n",
    "    \"fitness\",\n",
    "    0.000654452\n",
    "]\n",
    "\n",
    "# Extract numerical values from the list\n",
    "numerical_values = [x for x in input_values if isinstance(x, (float, int))]\n",
    "\n",
    "# Print the list of numerical values\n",
    "print(numerical_values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "cd913e4e-2e7b-4947-a367-0e6c62fe6e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "numerical_values_og = [\n",
    "    0.238523, 4.06938e-08, 0.00136106, 0.000307715, 0.000294039, 0.580586, \n",
    "    0.111827, 0.133209, 0.637912, 2.5717e-05, 2.36393e-16, 6.23198e-06, \n",
    "    5.51868e-17, 0.0189577, 0.00760604, 1.53508e-06, 2.2513e-05, 0.000196125, \n",
    "    0.0124593, 0.144752, 0.321249, 0.858058, 0.119282, 0.00320625, 0.000208478, \n",
    "    1.95253e-05, 6.36069e-06, 0.00054201, 0.991214, 0.86022, 0.000654452, \n",
    "    0.167066, 0.181709, 1.6127e-14\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "6a055afa-b847-4f3e-a640-3df792089046",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(numerical_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "684901b4-9acd-4b24-80a0-4262c185e406",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(numerical_values_og)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "ed9881b8-9ff2-4ced-a9e2-d513d7eb42f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(numerical_values) == set(numerical_values_og)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "103b3831-5bee-460b-b652-4edbabf50205",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False, False, False, False, False, False, False, False,  True,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False, False, False,\n",
       "       False, False, False, False, False, False, False])"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numerical_values_og == np.array(numerical_values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "cbb21856-4ea5-4901-9536-7e2d3187ebd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>2830555</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>119146348.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>ticks</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>1800</td></tr>\n",
       "                      <tr><td>Total Size</td><td>464.8 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 5<br/>file_version: 0.8<br/>model_type: nonWF<br/>name: sim<br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: late<br/>tick: 5<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>10154989</td>\n",
       "                      <td>309.9 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>900</td>\n",
       "                      <td>89.7 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>1.2 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>2133092</td>\n",
       "                      <td>77.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>2.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>9.6 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>0</td>\n",
       "                      <td>16 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7fcbe49279a0>"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tskit.load('results/arq_lowfreq_monogen_highbeta/moderatesel/optima0/subp0_tree_output.trees')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5578005-05a9-4c09-8408-3d1f9db67ff8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bec2f69f-cc6a-461a-93a2-a63a7f4c7677",
   "metadata": {},
   "outputs": [],
   "source": [
    "import snakemake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab60e216-b189-4b3c-a157-9ea9dcb5258a",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "module 'snakemake' has no attribute 'input'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_41914/2110642999.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0msnakemake\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'og_tree_offset'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: module 'snakemake' has no attribute 'input'"
     ]
    }
   ],
   "source": [
    "og_tree_offset = 'data/og_tree_offset.trees'\n",
    "print(og_tree_offset)\n",
    "mapper_realid_metadataid = 'mapper_realid_metadataid_wholegenomeoffset.csv'\n",
    "output_sim_tree ='results/arq_highfreq_twentypoly_lowbeta/moderatesel/optima4/subp4_tree_output.trees'\n",
    "print(output_sim_tree)\n",
    "output_sim_tree_wm = 'tretrees'\n",
    "print(output_sim_tree_wm)\n",
    "output_vcf = snakemake.output['output_vcf'] \n",
    "print(output_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29d5de0b-927f-4cad-8df5-af6a653a7675",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('reading')\n",
    "import pandas as pd\n",
    "import tskit\n",
    "import allel\n",
    "import random\n",
    "import numpy as np\n",
    "import tsinfer\n",
    "import pyslim\n",
    "\n",
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "print(og_tree_offset)\n",
    "mapper_realid_metadataid = snakemake.input['mapper_ids'] \n",
    "output_sim_tree = snakemake.input['output_sim_tree'] \n",
    "print(output_sim_tree)\n",
    "output_sim_tree_wm = snakemake.output['output_sim_tree_wm'] \n",
    "print(output_sim_tree_wm)\n",
    "output_vcf = snakemake.output['output_vcf'] \n",
    "print(output_vcf)\n",
    "\n",
    "def overlap_neutral_mut (ts_new, ts, mapper_realid_metadataid):\n",
    "    ## extract surviving ndoes and comapre them to our old ndoes to place mtuations in the right place\n",
    "    surviving_nodes = []\n",
    "    for i in ts_new.tables.nodes:\n",
    "        surviving_nodes.append(i.metadata['slim_id'])\n",
    "    ## new nodes id and the ids i gave them in the past\n",
    "    new_mapper = pd.DataFrame({'new_ids': range(0, len(ts_new.tables.nodes)), 'my_ids_metadata':surviving_nodes})\n",
    "    ## map old nodes with new nodes\n",
    "    mapper_lost_nodes = new_mapper.merge(mapper_realid_metadataid, left_on = 'my_ids_metadata', right_on = 'my_ids_metadata', how= 'right')\n",
    "\n",
    "    ## create a mask to only keep from the old nodes the ones that survived the simulation\n",
    "    mask = mapper_lost_nodes['new_ids'].notna()\n",
    "\n",
    "    tables_og = ts.dump_tables()\n",
    "\n",
    "    ## now filter old tables only based on surviving nodes \n",
    "    tables_og.nodes.replace_with(tables_og.nodes[mask])\n",
    "\n",
    "    ## now filter mutation table based on the surviving nodes, for that, extract the nodes \n",
    "    old_nodes = tables_og.mutations.node\n",
    "\n",
    "    old_nodes = pd.Series(old_nodes)\n",
    "\n",
    "    old_nodes.name = 'old_nodes'\n",
    "\n",
    "    ## create a dataframe relating the new and old nodes\n",
    "    replace_oldbynew_nodes = pd.merge(old_nodes, mapper_lost_nodes, left_on ='old_nodes', right_on = 'real_id', how= 'left')\n",
    "\n",
    "    ## create a mask to filter out all the mutations than has been lost \n",
    "    mask_mutations_lost = replace_oldbynew_nodes['new_ids'].notna()\n",
    "\n",
    "    ## filter out mutations that has been lost \n",
    "    table_mutations = tables_og.mutations[mask_mutations_lost]\n",
    "\n",
    "    ## now replace the old nodes ids by the new nodes ids with the mapper\n",
    "    ids_to_replace = replace_oldbynew_nodes.dropna()['new_ids']\n",
    "    table_mutations.node = np.array(ids_to_replace.astype('int32'))\n",
    "\n",
    "    ## and jsut set the sites from 0 to the length of mutation table \n",
    "    table_mutations.site = np.array(range(0, len(table_mutations))).astype('int32')\n",
    "\n",
    "    ## apply the same filter from the mutations table to the sites table \n",
    "    table_sites = tables_og.sites[mask_mutations_lost]  \n",
    "\n",
    "    ## now replace all this filter old tables in the new tree seq! \n",
    "    new_tables = ts_new.dump_tables()\n",
    "\n",
    "    new_tables.mutations.replace_with(table_mutations)\n",
    "\n",
    "    new_tables.sites.replace_with(table_sites)\n",
    "\n",
    "    ## make sure to compure mutations parents\n",
    "    new_tables.compute_mutation_parents()\n",
    "\n",
    "    ## create tree seq based on tables\n",
    "    tree_nm = new_tables.tree_sequence()\n",
    "\n",
    "    return tree_nm.simplify()\n",
    "\n",
    "def convert_tree_to_vcf (tree,name_vcf):\n",
    "    # create a vcf file from the treeseq \n",
    "    with open(name_vcf, 'w') as file:\n",
    "        # Pass the file object as the output parameter\n",
    "        tree.write_vcf(output=file)\n",
    "\n",
    "#import the old tree\n",
    "ts_old = tskit.load(og_tree_offset)\n",
    "#import mapper old nodes to new nodes\n",
    "mapper_realid_metadataid = pd.read_csv('../treeseq/mapper_realid_metadataid_wholegenome.csv')\n",
    "\n",
    "## ts new\n",
    "ts_new = tskit.load(output_sim_tree)\n",
    "\n",
    "ts_nm = overlap_neutral_mut(ts_new, ts_old, mapper_realid_metadataid)\n",
    "tskit.dump(ts_nm, output_sim_tree_wm)\n",
    "convert_tree_to_vcf(ts_nm, output_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea3291b9-11b4-4914-acda-02ccf30d9e2f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
