{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f697b9ef-748d-4397-a6fb-c8b916b15068",
   "metadata": {},
   "outputs": [],
   "source": [
    "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"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "27b9f9c2-1b6a-4a78-baf5-2a728f633d20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# once i have wholegenome_offset.trees\n",
    "import pandas as pd\n",
    "import tskit\n",
    "import allel\n",
    "import random\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import tsinfer\n",
    "import pyslim\n",
    "import os\n",
    "from collections import defaultdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e98412d-752b-4bda-8963-7260bbb96779",
   "metadata": {},
   "outputs": [],
   "source": [
    "## create th einitital python file before the simulations, to set up the architectures, create variances, and optimas "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd51083f-f215-4369-b9f3-20e45803d8d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c61571b-ae54-4c6e-9962-6de3a49f2683",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_vcf_offset = snakemake.input['og_vcf_offset'] \n",
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "\n",
    "pi_option =  snakemake.params['pi']\n",
    "pi = int(snakemake.params[pi_option])\n",
    "\n",
    "beta_option = snakemake.params['beta']\n",
    "beta = int(snakemake.params[beta_option])\n",
    "\n",
    "alelle_freq_option = snakemake.params['allele_freq']\n",
    "allele_freq = snakemake.params[alelle_freq_option]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "optima_qty = str(snakemake.params['optima_qty']) \n",
    "\n",
    "output_tree_seq_causalloci = snakemake.output[\"tree_seq_causalloci\"]\n",
    "output_loci_effectsize = snakemake.output[\"loci_effectsize\"]\n",
    "output_phenotypes = snakemake.output[\"phenotypes\"]\n",
    "output_optima_values = snakemake.output[\"optima_values\"]\n",
    "output_variance_values = snakemake.output[\"variance_values\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "e72137aa-16c8-4133-85eb-5e8250faa154",
   "metadata": {},
   "outputs": [],
   "source": [
    "## params hardcoded:\n",
    "og_vcf_offset = '../treeseq/wholegenome_offset.vcf'\n",
    "og_tree_offset = '../treeseq/wholegenome_offset_baselinetree.trees'\n",
    "\n",
    "pi = 20\n",
    "beta = 1\n",
    "\n",
    "allele_freq = [0.2, 0.3]\n",
    "lower_bound = float(allele_freq[0])\n",
    "upper_bound = float(allele_freq[1])\n",
    "\n",
    "\n",
    "#get the actual values\n",
    "optima_qty = 5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "1dc7d331-23db-412e-800d-46718b60fb72",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## functions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "125fdca8-815d-41b0-b2e9-6b53694c1214",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta):\n",
    "    alt_al_count = alt_al_per_pos.sum(axis=1)\n",
    "    alelle_dist = pd.DataFrame({'alt_al_count':alt_al_count, 'pos':pos})\n",
    "    alelle_dist['alt_al_freq'] = alelle_dist['alt_al_count'] / (n_ecotypes*2)\n",
    "    sim_freq_pos = alelle_dist[(alelle_dist['alt_al_freq'] < upper_bound) & (alelle_dist['alt_al_freq'] >= lower_bound)]['pos']\n",
    "    selected_sites = sim_freq_pos.sample(pi).values\n",
    "    sc = np.random.normal(0, beta, pi)\n",
    "    pos_sc = pd.DataFrame({'pos': selected_sites, 'sc': sc})\n",
    "    return pos_sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "b9f0cfa3-2dd2-48ac-86a4-8ef70d3d6e05",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_phenotypes(pos,pos_sc, alt_al_per_pos):\n",
    "    mask_positions = pd.Series(pos).isin(pos_sc['pos'])\n",
    "    alt_al_per_pos_selected_sites = alt_al_per_pos[mask_positions]\n",
    "    phenotypes = []\n",
    "    for i in range(alt_al_per_pos_selected_sites.shape[1]):\n",
    "        gen_effectsize = np.multiply(alt_al_per_pos_selected_sites[:, i] , pos_sc['sc'])\n",
    "        phenotypes.append(gen_effectsize.sum())\n",
    "    return phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "981b17fc-83c9-4581-96ab-4cc6d287d530",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_optima(phenotypes):\n",
    "    max_pheno = max(phenotypes)\n",
    "    min_pheno = min(phenotypes)\n",
    "\n",
    "    length = max_pheno - min_pheno\n",
    "    step = length/(int(optima_qty) - 1)\n",
    "    optima = [round(min_pheno + i * step, 4) for i in range(0, int(optima_qty))]\n",
    "    return optima"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "17bf89d2-0352-4a9f-917d-977679d7d945",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_variances(phenotypes, optima):\n",
    "    range_pheno =  max(phenotypes) - min(phenotypes)\n",
    "    dist_between_env = range_pheno / len(optima)\n",
    "    ## Strong selection, fitness 0 in the adyacent environemnt\n",
    "    sd1 = dist_between_env / 3\n",
    "    variance1 = sd1**2\n",
    "    ## moderate selection, fitness 0 in the other extremee environemnt \n",
    "    sd2 = (dist_between_env * 4) / 3 ## 3 sd will be in between 4  environments \n",
    "    variance2 = sd2**2\n",
    "    ## weak selection, half fitness in the other extreme environment \n",
    "    sd3 = (dist_between_env * 8) / 3\n",
    "    variance3 = sd3**2\n",
    "    variances = [sd1, sd2, sd3]\n",
    "    return variances\n",
    "\n",
    "def keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc):\n",
    "    ts = tskit.load(og_tree_offset)\n",
    "\n",
    "    # dumpt the tables from the tree\n",
    "    tables = ts.dump_tables()\n",
    "\n",
    "    ## extract all teh sites from the og tree\n",
    "    complete_sites = pd.Series(tables.sites.position)\n",
    "\n",
    "    # create a mask to filter only the ones present in the selected sites (causal loci)\n",
    "    mask_delete_sites = complete_sites.isin(pos_sc['pos'])\n",
    "\n",
    "    ## replace the table only with the causal sites, and same for mutation tables\n",
    "    tables.sites.replace_with(tables.sites[mask_delete_sites])\n",
    "    tables.mutations.replace_with(tables.mutations[mask_delete_sites])\n",
    "    ## extract the new site index\n",
    "    tables.mutations.site = np.array(range(0, len(tables.mutations))).astype('int32')\n",
    "\n",
    "    ## ge tthe positions and sc in teh right order \n",
    "    pos_table = pd.Series(tables.sites.position).reset_index()\n",
    "    right_order_pos = pos_sc.merge(pos_table, left_on='pos',right_on =0).sort_values('index')\n",
    "\n",
    "    ## create the tree to then modify it \n",
    "    new_ts = tables.tree_sequence()\n",
    "    tables = new_ts.dump_tables()\n",
    "\n",
    "    ## chance the ancestral state to empty or slim will complain\n",
    "    tables.sites.clear()\n",
    "    for s in new_ts.sites():\n",
    "        tables.sites.append(s.replace(ancestral_state=\"\"))\n",
    "\n",
    "    ## add the selection coefficient and the rigth emtadata fro slim \n",
    "    tables.mutations.clear()\n",
    "    for k, (m, sc) in enumerate(zip(new_ts.mutations(), right_order_pos['sc'])):\n",
    "        mm = pyslim.default_slim_metadata('mutation_list_entry')\n",
    "        mm['selection_coeff'] = sc\n",
    "        tables.mutations.append(\n",
    "            m.replace(derived_state=str(k), metadata={'mutation_list': [mm]}))\n",
    "        \n",
    "    return tables.tree_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "1e90d1e9-9733-4119-b48b-1ae17d2e3b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "## for this im gonna use the og vcf file wth the offset to be able to map the positions correctly \n",
    "vcf_og = allel.read_vcf(og_vcf_offset, fields=[\"calldata/GT\", 'variants/POS' , 'samples'])\n",
    "geno_og = vcf_og[\"calldata/GT\"]\n",
    "samples = vcf_og['samples']\n",
    "pos = vcf_og['variants/POS']\n",
    "\n",
    "n_ecotypes = len(vcf_og['samples'])\n",
    "alt_al_per_pos = geno_og.sum(axis=2) \n",
    "\n",
    "pos_sc = calc_pos_sc(alt_al_per_pos, pos, n_ecotypes, allele_freq, pi, beta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16706eb9-38fb-4d1f-8a5d-7714811dcc48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "f1996aab-1c7c-42d5-b351-c33a9a1d6aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "phenotypes = calc_phenotypes(pos,pos_sc, alt_al_per_pos)\n",
    "\n",
    "optima = calc_optima(phenotypes)\n",
    "\n",
    "variances = calc_variances(phenotypes, optima)\n",
    "\n",
    "variances\n",
    "\n",
    "pd.DataFrame(index =  ['strongsel','moderatesel','lowsel'],data = { 'var': variances}).to_csv('sel_var.csv')\n",
    "\n",
    "pd.DataFrame(data = {'selection': ['strongsel','moderatesel','lowsel'], 'var': variances}).to_csv('sel_var.csv')\n",
    "\n",
    "## save \n",
    "\n",
    "pd.Series(phenotypes).to_csv(output_phenotypes)\n",
    "pos_sc.to_csv(output_loci_effectsize)\n",
    "\n",
    "with open(output_optima_values, 'w') as file:\n",
    "    for element in optima:\n",
    "        file.write(str(element) + '\\n')  # Write element followed by a newline\n",
    "\n",
    "with open(output_variance_values, 'w') as file:\n",
    "    for element in variances:\n",
    "        file.write(str(round(element,4)) + '\\n')  # Write element followed by a newline\n",
    "\n",
    "### filter tree\n",
    "\n",
    "pre_slim_tree = keep_only_causal_sites_and_mutations(og_tree_offset, pos_sc)\n",
    "\n",
    "## save tree\n",
    "\n",
    "pre_slim_tree.dump(output_tree_seq_causalloci)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef1d73bd-9371-415b-9ee9-327506d4d64f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e6547a54-643f-4392-b93a-728f693b01f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## tree postprocesssing "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f304969-0805-4a21-946a-83bcc370cfc8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "63c188e0-249a-4975-81ae-8ac0fbfb0a0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset = snakemake.input['og_tree_offset'] \n",
    "mapper_realid_metadataid = snakemake.input['mapper_ids'] \n",
    "output_sim_tree = snakemake.input['output_sim_tree'] \n",
    "output_sim_tree_wm = snakemake.output['output_sim_tree_wm'] \\\n",
    "output_vcf = snakemake.output['output_vcf'] \n",
    "\n",
    "def overlap_neutral_mut (ts_new, ts, mapper_realid_metadataid):\n",
    "    ## extract surviving ndoes and comapre them to our old ndoes to place mtuations in the right place\n",
    "    surviving_nodes = []\n",
    "    for i in ts_new.tables.nodes:\n",
    "        surviving_nodes.append(i.metadata['slim_id'])\n",
    "    ## new nodes id and the ids i gave them in the past\n",
    "    new_mapper = pd.DataFrame({'new_ids': range(0, len(ts_new.tables.nodes)), 'my_ids_metadata':surviving_nodes})\n",
    "    ## map old nodes with new nodes\n",
    "    mapper_lost_nodes = new_mapper.merge(mapper_realid_metadataid, left_on = 'my_ids_metadata', right_on = 'my_ids_metadata', how= 'right')\n",
    "\n",
    "    ## create a mask to only keep from the old nodes the ones that survived the simulation\n",
    "    mask = mapper_lost_nodes['new_ids'].notna()\n",
    "\n",
    "    tables_og = ts.dump_tables()\n",
    "\n",
    "    ## now filter old tables only based on surviving nodes \n",
    "    tables_og.nodes.replace_with(tables_og.nodes[mask])\n",
    "\n",
    "    ## now filter mutation table based on the surviving nodes, for that, extract the nodes \n",
    "    old_nodes = tables_og.mutations.node\n",
    "\n",
    "    old_nodes = pd.Series(old_nodes)\n",
    "\n",
    "    old_nodes.name = 'old_nodes'\n",
    "\n",
    "    ## create a dataframe relating the new and old nodes\n",
    "    replace_oldbynew_nodes = pd.merge(old_nodes, mapper_lost_nodes, left_on ='old_nodes', right_on = 'real_id', how= 'left')\n",
    "\n",
    "    ## create a mask to filter out all the mutations than has been lost \n",
    "    mask_mutations_lost = replace_oldbynew_nodes['new_ids'].notna()\n",
    "\n",
    "    ## filter out mutations that has been lost \n",
    "    table_mutations = tables_og.mutations[mask_mutations_lost]\n",
    "\n",
    "    ## now replace the old nodes ids by the new nodes ids with the mapper\n",
    "    ids_to_replace = replace_oldbynew_nodes.dropna()['new_ids']\n",
    "    table_mutations.node = np.array(ids_to_replace.astype('int32'))\n",
    "\n",
    "    ## and jsut set the sites from 0 to the length of mutation table \n",
    "    table_mutations.site = np.array(range(0, len(table_mutations))).astype('int32')\n",
    "\n",
    "    ## apply the same filter from the mutations table to the sites table \n",
    "    table_sites = tables_og.sites[mask_mutations_lost]  \n",
    "\n",
    "    ## now replace all this filter old tables in the new tree seq! \n",
    "    new_tables = ts_new.dump_tables()\n",
    "\n",
    "    new_tables.mutations.replace_with(table_mutations)\n",
    "\n",
    "    new_tables.sites.replace_with(table_sites)\n",
    "\n",
    "    ## make sure to compure mutations parents\n",
    "    new_tables.compute_mutation_parents()\n",
    "\n",
    "    ## create tree seq based on tables\n",
    "    tree_nm = new_tables.tree_sequence()\n",
    "\n",
    "    return tree_nm.simplify()\n",
    "\n",
    "def convert_tree_to_vcf (tree,name_vcf):\n",
    "    # create a vcf file from the treeseq \n",
    "    with open(name_vcf, 'w') as file:\n",
    "        # Pass the file object as the output parameter\n",
    "        tree.write_vcf(output=file)\n",
    "\n",
    "#import the old tree\n",
    "ts_old = tskit.load(og_tree_offset)\n",
    "#import mapper old nodes to new nodes\n",
    "mapper_realid_metadataid = pd.read_csv(mapper_realid_metadataid)\n",
    "\n",
    "## ts new\n",
    "\n",
    "if os.path.exists(output_sim_tree) and os.path.getsize(output_sim_tree) <= 1:\n",
    "    print('empty_tree')\n",
    "    with open(output_vcf, \"w\"):\n",
    "        pass  # Create an empty vcf file \n",
    "    with open(output_sim_tree_wm, \"w\"):\n",
    "        pass  # Create an empty tree file \n",
    "elif os.path.exists(output_sim_tree) and os.path.getsize(output_sim_tree) > 1:\n",
    "    ts_new = tskit.load(output_sim_tree)\n",
    "    ts_nm = overlap_neutral_mut(ts_new, ts_old, mapper_realid_metadataid)\n",
    "    ts_nm.dump(output_sim_tree_wm)\n",
    "    convert_tree_to_vcf(ts_nm, output_vcf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93b1244e-8a50-4c97-a1e6-9aa6feb41844",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c478a4b-5c19-4458-bfee-4eb876cdd80b",
   "metadata": {},
   "outputs": [],
   "source": [
    "    ts = tskit.load(og_tree_offset)\n",
    "\n",
    "    # dumpt the tables from the tree\n",
    "    tables = ts.dump_tables()\n",
    "\n",
    "    ## extract all teh sites from the og tree\n",
    "    complete_sites = pd.Series(tables.sites.position)\n",
    "\n",
    "    # create a mask to filter only the ones present in the selected sites (causal loci)\n",
    "    mask_delete_sites = complete_sites.isin(pos_sc['pos'])\n",
    "\n",
    "    ## replace the table only with the causal sites, and same for mutation tables\n",
    "    tables.sites.replace_with(tables.sites[mask_delete_sites])\n",
    "    tables.mutations.replace_with(tables.mutations[mask_delete_sites])\n",
    "    ## extract the new site index\n",
    "    tables.mutations.site = np.array(range(0, len(tables.mutations))).astype('int32')\n",
    "\n",
    "    ## ge tthe positions and sc in teh right order \n",
    "    pos_table = pd.Series(tables.sites.position).reset_index()\n",
    "    right_order_pos = pos_sc.merge(pos_table, left_on='pos',right_on =0).sort_values('index')\n",
    "\n",
    "    ## create the tree to then modify it \n",
    "    new_ts = tables.tree_sequence()\n",
    "    tables = new_ts.dump_tables()\n",
    "\n",
    "    ## chance the ancestral state to empty or slim will complain\n",
    "    tables.sites.clear()\n",
    "    for s in new_ts.sites():\n",
    "        tables.sites.append(s.replace(ancestral_state=\"\"))\n",
    "\n",
    "    ## add the selection coefficient and the rigth emtadata fro slim \n",
    "    tables.mutations.clear()\n",
    "    for k, (m, sc) in enumerate(zip(new_ts.mutations(), right_order_pos['sc'])):\n",
    "        mm = pyslim.default_slim_metadata('mutation_list_entry')\n",
    "        mm['selection_coeff'] = sc\n",
    "        tables.mutations.append(\n",
    "            m.replace(derived_state=str(k), metadata={'mutation_list': [mm]}))\n",
    "        \n",
    "    return tables.tree_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05bee502-f70d-4fe8-97b7-a349f2e0fb75",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9e5c8440-93ac-4c74-abc4-9666356d85db",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/home/tbellagio/scratch/slim_grenenet/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "65832e70-404b-4418-8bdf-0ac678f29477",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyslim\n",
    "import tskit\n",
    "\n",
    "baseline_tree = tskit.load('og_tree_offset.trees')\n",
    "\n",
    "tables = baseline_tree.dump_tables()\n",
    "\n",
    "tables.sites.clear()\n",
    "for s in baseline_tree.sites():\n",
    "    tables.sites.append(s.replace(ancestral_state=\"\"))\n",
    "\n",
    "tables.mutations.clear()\n",
    "mm = pyslim.default_slim_metadata('mutation_list_entry')\n",
    "for k, m in enumerate(baseline_tree.mutations()):\n",
    "    tables.mutations.append(\n",
    "        m.replace(derived_state=str(k), metadata={'mutation_list': [mm]}))\n",
    "\n",
    "og_tree_offset_for_slim = tables.tree_sequence()\n",
    "\n",
    "og_tree_offset_for_slim.dump('og_tree_offset_for_slim.trees')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "547e4718-eee1-4018-af5e-b21bb0a52781",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "add0f9a2-5947-49ea-a0be-4a63a1999036",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree = tskit.load(path + 'data/og_tree_offset_for_slim.trees')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0a05e3b9-b9fc-417f-bad7-674a9c07f8e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1674685</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>119146348.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>uncalibrated</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>462</td></tr>\n",
       "                      <tr><td>Total Size</td><td>758.0 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 1<br/>description: <br/>file_version: 0.8<br/>model_type: nonWF<br/>name: <br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: early<br/>tick: 1<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>8960512</td>\n",
       "                      <td>273.5 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>231</td>\n",
       "                      <td>22.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>184.1 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>2146164</td>\n",
       "                      <td>77.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>2.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>1</td>\n",
       "                      <td>587 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>154.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7ff6c799f310>"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9d9d74da-cfdf-4072-99ae-356a51235508",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1622022</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>119146348.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>uncalibrated</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>462</td></tr>\n",
       "                      <tr><td>Total Size</td><td>862.3 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 1<br/>description: <br/>file_version: 0.8<br/>model_type: nonWF<br/>name: <br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: early<br/>tick: 1<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>11929901</td>\n",
       "                      <td>364.1 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>231</td>\n",
       "                      <td>22.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>184.1 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>1899499</td>\n",
       "                      <td>68.8 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>2.5 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>2</td>\n",
       "                      <td>1.1 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>154.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7ff65bfa46a0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tree.simplify()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d34e159f-5556-4df1-a838-02f256b6d0ce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5d3b4359-a5a6-4d87-8e22-5137ab53c23f",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset_after_cloning = tskit.load(path + 'data/og_tree_offset_after_cloning.trees')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c9be4b6c-8132-4a32-b932-684bc2441e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset_after_cloning = og_tree_offset_after_cloning.simplify()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "61bda30d-901a-4f72-9bc1-bb1a5e4e6321",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_tree_to_vcf (tree,name_vcf):\n",
    "    # create a vcf file from the treeseq \n",
    "    with open(name_vcf, 'w') as file:\n",
    "        # Pass the file object as the output parameter\n",
    "        tree.write_vcf(output=file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "dc1280ec-d4ed-441a-9270-fc9181107404",
   "metadata": {},
   "outputs": [],
   "source": [
    "convert_tree_to_vcf(og_tree_offset_after_cloning, path + 'data/wholegenome_offset_after_cloning.vcf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfdb7bf5-0772-4d57-8204-d18e30a5082e",
   "metadata": {},
   "outputs": [],
   "source": [
    "RUN THE CORECTION FOR POSITIONS "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4b24ddb-3215-4437-9a94-6b4dbb1889c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "og_tree_offset_after_cloning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a585c3-87f4-4e7c-90d6-c202e38dbd30",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c6332897-b52e-43f9-9978-4cf94be9ed95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div>\n",
       "              <style>\n",
       "                .tskit-table thead tr th {text-align: left;padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td {padding: 0.5em 0.5em;}\n",
       "                .tskit-table tbody tr td:first-of-type {text-align: left;}\n",
       "                .tskit-details-label {vertical-align: top; padding-right:5px;}\n",
       "                .tskit-table-set {display: inline-flex;flex-wrap: wrap;margin: -12px 0 0 -12px;width: calc(100% + 12px);}\n",
       "                .tskit-table-set-table {margin: 12px 0 0 12px;}\n",
       "                details {display: inline-block;}\n",
       "                summary {cursor: pointer; outline: 0; display: list-item;}\n",
       "              </style>\n",
       "              <div class=\"tskit-table-set\">\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"padding:0;line-height:21px;\">\n",
       "                          <img style=\"height: 32px;display: inline-block;padding: 3px 5px 3px 0;\" src=\"https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg\"/>\n",
       "                          <a target=\"_blank\" href=\"https://tskit.dev/tskit/docs/latest/python-api.html#the-treesequence-class\"> Tree Sequence </a>\n",
       "                        </th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                      <tr><td>Trees</td><td>1622022</td></tr>\n",
       "                      <tr><td>Sequence Length</td><td>119146348.0</td></tr>\n",
       "                      <tr><td>Time Units</td><td>ticks</td></tr>\n",
       "                      <tr><td>Sample Nodes</td><td>5082</td></tr>\n",
       "                      <tr><td>Total Size</td><td>862.9 MiB</td></tr>\n",
       "                      <tr>\n",
       "                        <td>Metadata</td><td style=\"text-align: left;\">\n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\"></span>\n",
       "                  <details open>\n",
       "                    <summary>dict</summary>\n",
       "                    \n",
       "                <div>\n",
       "                  <span class=\"tskit-details-label\">SLiM:</span>\n",
       "                  <details >\n",
       "                    <summary>dict</summary>\n",
       "                    cycle: 1<br/>file_version: 0.8<br/>model_type: nonWF<br/>name: sim<br/>nucleotide_based: False<br/>separate_sexes: False<br/>spatial_dimensionality: <br/>spatial_periodicity: <br/>stage: late<br/>tick: 1<br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                <br/>\n",
       "                  </details>\n",
       "                </div>\n",
       "                </td></tr>\n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "                <div class=\"tskit-table-set-table\">\n",
       "                  <table class=\"tskit-table\">\n",
       "                    <thead>\n",
       "                      <tr>\n",
       "                        <th style=\"line-height:21px;\">Table</th>\n",
       "                        <th>Rows</th>\n",
       "                        <th>Size</th>\n",
       "                        <th>Has Metadata</th>\n",
       "                      </tr>\n",
       "                    </thead>\n",
       "                    <tbody>\n",
       "                    \n",
       "                  <tr>\n",
       "                    <td>Edges</td>\n",
       "                      <td>11934521</td>\n",
       "                      <td>364.2 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Individuals</td>\n",
       "                      <td>2541</td>\n",
       "                      <td>248.1 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Migrations</td>\n",
       "                      <td>0</td>\n",
       "                      <td>8 Bytes</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Mutations</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>184.1 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Nodes</td>\n",
       "                      <td>1904119</td>\n",
       "                      <td>69.0 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Populations</td>\n",
       "                      <td>1</td>\n",
       "                      <td>2.3 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Provenances</td>\n",
       "                      <td>3</td>\n",
       "                      <td>4.4 KiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        \n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                  <tr>\n",
       "                    <td>Sites</td>\n",
       "                      <td>3235480</td>\n",
       "                      <td>154.3 MiB</td>\n",
       "                      <td style=\"text-align: center;\">\n",
       "                        ✅\n",
       "                      </td>\n",
       "                    </tr>\n",
       "                \n",
       "                    </tbody>\n",
       "                  </table>\n",
       "                </div>\n",
       "              </div>\n",
       "            </div>\n",
       "            "
      ],
      "text/plain": [
       "<tskit.trees.TreeSequence at 0x7ff6c7921f70>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "og_tree_offset_after_cloning.simplify()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e0df5eb-e5c5-4f12-8961-cbf1f1848a18",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (snakemake_pipeline)",
   "language": "python",
   "name": "pipeline_snakemake"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
